CAS-399 - Add nav list to global navigation component

aliuk2012 opened the pull request:
# Context

https://dsdmoj.atlassian.net/browse/CAS-399

<!-- Is there a Trello ticket you can link to? -->
<!-- Do you need to add any environment variables? -->
<!-- Is an ADR required? An ADR should be added if this PR introduces a change to the architecture. -->

# Changes in this PR
Any user who has `assessor` role will now see a list of links in the primary navigation. If they do not `assessor` role then they will not see any links in the primary navigation.

As part of this PR, there were changes to the bookings index page instead of `/bookings/{status}` we now have a `/bookings` that will redirect the user to the first page of bookings by default.

The url to access the reports page also changed from `/reports/bookings/` to `/reports`

## Screenshots of UI changes

### Before

### After

# Release checklist

[Release process documentation](https://dsdmoj.atlassian.net/wiki/spaces/AP/pages/edit-v2/4247847062?draftShareId=a1c360ab-bd31-4db1-aae3-7cc002761de9).

## Pre-merge checklist

- [ ] Are any changes required to dependent services for this change to work?
  (eg. CAS API)
- [ ] Have they been released to production already?

## Post-merge checklist

- [ ] [Manually approve](https://dsdmoj.atlassian.net/wiki/spaces/AP/pages/4247847062/Release+process#Manual-releases)
  release to test
- [ ] [Manually approve](https://dsdmoj.atlassian.net/wiki/spaces/AP/pages/4247847062/Release+process#Manual-releases)
  release to preprod
- [ ] [Manually approve](https://dsdmoj.atlassian.net/wiki/spaces/AP/pages/4247847062/Release+process#Manual-releases)
  release to prod

<!-- Should a release fail at any step, you as the author should now lead the work to
fix it as soon as possible. You can monitor deployment failures in CircleCI
itself and application errors are found in
[Sentry](https://ministryofjustice.sentry.io/issues/?project=4504129156218880&referrer=sidebar&statsPeriod=24h).
Both events should be automatically sent to our [Slack
channel](https://mojdt.slack.com/archives/C048BJS7S2F). -->

aliuk2012 committed:
Stub out new userService to set a list of navigation links

aliuk2012 committed:
Pass users.primaryNavigationList to moj primary navigation component

aliuk2012 committed:
Update getActingUserPrimaryNavigationList to accept user and currentpage

Based on the users role, if they are an assessor they will get a list of links that they can visit. Otherwise, and empty array is returned.

Depending on the currentpage the user is viewing, if it contains part of the url that the nav link is then that nav link will display as being active.

aliuk2012 committed:
Add a root for Bookings index and then redirect to the existing provisional bookings page

We trying to bring all bookings under `/bookings` path. this helps with marking the primary nav list item. Its not perfect but it will do for the time being

aliuk2012 committed:
Add a root for Reports index and then redirect to the existing report bookings page

We trying to bring reports under `/reports` path instead of `/reports/bookings`. this helps with marking the primary nav list item.

froddd commented on e2e/tests/stepDefinitions/manage/bookingReport.ts:
```diff
@@ -3,7 +3,7 @@ import { Given, Then } from '@badeball/cypress-cucumber-preprocessor'
 import Page from '../../../../cypress_shared/pages/page'
 import DashboardPage from '../../../../cypress_shared/pages/temporary-accommodation/dashboardPage'
-import ReportNewPage from '../../../../cypress_shared/pages/temporary-accommodation/manage/reportNew'
+import ReportNewPage from '../../../../cypress_shared/pages/temporary-accommodation/manage/reportIndex'
```
Can we get this one renamed too? Obviously will also need other usages in this file done also!

```suggestion
import ReportIndexPage from '../../../../cypress_shared/pages/temporary-accommodation/manage/reportIndex'
```

froddd commented on server/controllers/temporary-accommodation/manage/reportsController.ts:
```diff
@@ -34,6 +34,12 @@ export default class ReportsController {
+    return async (req: Request, res: Response) => {
+      return res.redirect(301, paths.reports.index({}))
+    }
+  }
```
It may be nice to think of a way of flagging this as deprecated -- maybe renaming the method would be sufficient, or creating a generic decorator that would handle the redirect? It seems we'll be needing this in a couple of places, and it would help any new dev understand the reason behind the redirect.

aliuk2012 replied:
I'm thinking we drop this completely in favour or doing the redirect directly in the routes file. Any thoughts?

froddd replied:
Yes, that seems like the best option -- less code when new redirects are needed, and the concern of routing is all kept in one place. Better!

froddd commented on server/middleware/populateCurrentUser.ts:
```diff
@@ -13,6 +13,9 @@ export default function populateCurrentUser(userService: UserService): RequestHa
         req.session.userDetails = userDetails
         req.session.probationRegion = userDetails.region
 
+        // add nav here
```
This comment doesn't add much, can be removed.

froddd commented on server/routes/temporary-accommodation/manage.ts:
```diff
@@ -252,7 +252,7 @@ export default function routes(controllers: Controllers, services: Services, rou
   })
 
-  get(paths.reports.new.pattern, reportsController.new(), { auditEvent: 'VIEW_REPORT_CREATE' })
+  get(paths.reports.index(), reportsController.index(), { auditEvent: 'VIEW_REPORT_INDEX' })
```
We still need a reference to the current `paths.reports.new()` here, otherwise it won't be handled at all (as it stands trying to load `/reports/bookings` returns an error)

In fact this may be a better place to hold the redirect itself, with a new `redirectsController` maybe? This would remove the need to add the redirect logic everywhere it is needed, and possibly make all redirects more discoverable, and easier to implement when needed.

```suggestion
  // Redirect using a generic redirects controller, with redirect method taking location and optional status code
  get(paths.reports.new(), redirectsController.redirect(paths.reports.index(), 301), { auditEvent: 'REDIRECT_REPORT_NEW' })
  get(paths.reports.index(), reportsController.index(), { auditEvent: 'VIEW_REPORT_INDEX' })
```

aliuk2012 replied:
Love this suggestion, I was following the existing pattern that was used for Referralsl. I'll update referrals as well so theres no redirects happening in the controllers

froddd commented on server/routes/temporary-accommodation/manage.ts:
```diff
@@ -302,6 +302,10 @@ export default function routes(controllers: Controllers, services: Services, rou
 
+  get(paths.bookings.index(), bookingsController.index(), { auditEvent: 'VIEW_BOOKINGS_INDEX' })
+
+  get(paths.bookings.index(), bookingsController.index(), { auditEvent: 'VIEW_BOOKINGS_INDEX' })
```
Unnecessary duplication?

Could also be handled by a generic redirects controller as mentioned earlier?

froddd commented on server/services/userService.test.ts:
```diff
@@ -52,4 +54,42 @@ describe('User service', () => {
+      })
+
+      describe('when a user is viewing a nav list item page', () => {
+        it.each(navigationList)(`returns active when a user is viewing %p page with %p link`, ({ text, href }) => {
```
Could this also test the correct behaviour with paths that only _contain_ the nav url? I.e. showing 'Bookings' as active for `/bookings/foo/bar`

froddd commented on server/services/userService.ts:
```diff
@@ -30,4 +33,28 @@ export default class UserService {
+        }
+        return true
+      })
+      return navList
```
This feels odd -- `every` is typically used to assert every element meets a condition. Could this use `map` instead?

```suggestion
      return navList.map(item => ({
        ...item,
        active: currentPage.includes(item.href) || undefined, // or leave this as `false` instead of undefined
      }))
```

froddd reviewed (commented):
Some suggestion regarding handling of redirects to make them more discoverable.

It would also be good to have some integration test for the nav, to show that it is actually getting rendered correctly depending on the user. There are some integration tests around roles already, maybe add to that?

aliuk2012 committed:
Create a new generic RedirectController for routes

Instead of having loads of Controllers with methods that redirect to another url, it would be better to do all these redirects directly in routes file

aliuk2012 committed:
Centralise redirects to be routes

Previously, we had a couple controller methods created which were purely for redirecting to another page. This is a code smell and could be done much easier by using a generic redirect controller.

We now have this controller in place so this commit is to update the previous pattern to now all be redirected straight from routes and not in the individual controllers.

aliuk2012 committed:
Add extra tests for getActingUserPrimaryNavigationList

Other url which the user may visit and which link we would expect to be active based on the url

froddd reviewed (approved):
Perfect -- I think moving to the RedirectsController approach works really well. Keeps concern in one place which is really nice!
